server/graphql/general/src/queries/temperature_chart.rs

```diff
@@ -0,0 +1,391 @@
+use std::collections::HashMap;
+
+use async_graphql::{dataloader::DataLoader, *};
+use chrono::{DateTime, NaiveDateTime, Utc};
+use graphql_core::{
+    loader::SensorByIdLoader,
+    standard_graphql_error::{validate_auth, StandardGraphqlError},
+    ContextExt,
+};
+use graphql_types::types::{SensorNode, TemperatureLogFilterInput};
+use repository::{TemperatureChartRow, TemperatureLogFilter};
+use service::{
+    auth::{Resource, ResourceAccessRequest},
+    temperature_chart::{
+        TemperatureChart, TemperatureChartError as ServiceError, TemperatureChartInput,
+    },
+};
+
+#[derive(Union)]
+pub enum TemperatureChartResponse {
+    Response(TemperatureChartNode),
+}
+
+pub fn temperature_chart(
+    ctx: &Context<'_>,
+    store_id: String,
+    from_datetime: DateTime<Utc>,
+    to_datetime: DateTime<Utc>,
+    number_of_data_points: i32,
+    filter: Option<TemperatureLogFilterInput>,
+) -> Result<TemperatureChartResponse> {
+    let user = validate_auth(
+        ctx,
+        &ResourceAccessRequest {
+            resource: Resource::QueryTemperatureLog,
+            store_id: Some(store_id.clone()),
+        },
+    )?;
+
+    let service_provider = ctx.service_provider();
+    let service_context = service_provider.context(store_id.clone(), user.user_id)?;
+
+    let temperature_chart = service_provider
+        .temperature_chart_service
+        .get_temperature_chart(
+            &service_context,
+            TemperatureChartInput {
+                from_datetime: from_datetime.naive_utc(),
+                to_datetime: to_datetime.naive_utc(),
+                number_of_data_points: number_of_data_points,
+                filter: filter.map(TemperatureLogFilter::from),
+            },
+        )
+        .map_err(map_error)?;
+
+    Ok(TemperatureChartResponse::Response(
+        TemperatureChartNode::from_domain(temperature_chart)?,
+    ))
+}
+
+fn map_error(error: ServiceError) -> async_graphql::Error {
+    use StandardGraphqlError::*;
+    let formatted_error = format!("{:#?}", error);
+
+    let graphql_error = match error {
+        ServiceError::TooManyDataPoints
+        | ServiceError::AtLeastThreeDataPoint
+        | ServiceError::ToDateTimeMustBeAfterFromDatetime => BadUserInput(formatted_error),
+        ServiceError::DatabaseError(_) => InternalError(formatted_error),
+    };
+
+    graphql_error.extend()
+}
+
+#[derive(Default, SimpleObject)]
+pub struct TemperatureChartNode {
+    sensors: Vec<SensorAxisNode>,
+}
+
+impl TemperatureChartNode {
+    fn from_domain(
```

Comment:
You can check out tests for example of what is being mapped to what, mainly it's just creating an array of sensors, each sensor has data points, with any points missing from intervals being added added as 'empty' point

Reply:
I was weighing in 'where' to do this mapping, although it seems more service like task, but it's so much easier to test in steps, since service methods can be overwritten in test, it's seems easier to test the pure mapping, rather then inserting data in DB and testing how it maps in service (since then need to also think about what repo is doing with the raw data)

Reply:
If repo's were also service like and their result could be simulated, then probably would have gone with doing this mapping in service